uk.co.compucorp.civicrm.hrleaveandabsences/js/src/leave-absences/leave-type-wizard/form/components/leave-type-wizard-form.component.js

```diff
@@ -12,7 +14,165 @@ define([], function () {
+              lineBreak,
+              'If you enable this option the leave type will simply be given the title "Leave".',
+              'Managers and administrators will continue to see the actual leave title of the request.',
+              lineBreak,
+              'This may be helpful if the leave type title is sensitive - i.e. Unpaid leave or Compassionate leave.'
+            ].join('')
+          }
+        ]
+      },
+      {
+        name: 'leave-requests',
+        label: 'Leave Requests',
+        fields: [
+          {
+            name: 'max_consecutive_leave_days',
+            label: 'Max consecutive duration (Leave blank for unlimited)',
+            helpText: [
+              'Configure the maximum duration of consecutive leave permitted to be selected in a single leave request.',
+              'You can leave this field blank for unlimited duration of leave.'
+            ].join('')
+          }
+        ]
+      },
+      {
+        name: 'public-holidays',
+        label: 'Public Holidays',
+        fields: [
+          {
+            name: 'must_take_public_holiday_as_leave',
+            label: 'Do staff work on public holidays?'
+          }
+        ]
+      },
+      {
+        name: 'carry-forwards',
+        label: 'Carry Forwards',
+        fields: [
+          {
+            name: 'allow_carry_forward',
+            label: 'Allow carry forward?'
+          }
+        ]
+      }
+    ];
+
+    vm.$onInit = $onInit;
+    vm.getSettingsTabFields = getSettingsTabFields;
+    vm.openSection = openSection;
+    vm.openSettingsTab = openSettingsTab;
+    vm.selectLeaveTypeCategory = selectLeaveTypeCategory;
+
+    function $onInit () {
+      initDefaultView();
+    }
+
+    /**
+     * Gets fields for the currently active settings tab
+     *
+     * @return {Array} collection of fields
+     */
+    function getSettingsTabFields () {
+      var activeTab = _.find(vm.settingsTabs, { active: true });
+
+      return activeTab.fields;
+    }
+
+    /**
+     * Initiates the default view:
+     * - selects Leave category;
+     * - expands the General section and leaves the Settings section collapsed;
+     * - selects Basic Details settings tab.
+     */
+    function initDefaultView () {
+      vm.leaveTypeCategory = 'leave';
+
+      openSection('general');
+      openSettingsTab('basic-details');
+    }
+
+    /**
+     * Opens a section by its name and collapses all other sections
+     *
+     * @param {String} sectionName
+     */
+    function openSection (sectionName) {
+      vm.sections.forEach(function (section) {
+        section.active = false;
+      });
+
+      _.find(vm.sections, { name: sectionName }).active = true;
+    }
+
+    /**
+     * Opens a settings tab by its name and collapses all other settings tabs
+     *
+     * @param {String} settingsTabName
+     */
+    function openSettingsTab (settingsTabName) {
```

Comment:
I just realized, aren't you reinventing the wheel? [`ui.bootstrap`](http://angular-ui.github.io/bootstrap/#!#tabs) already gives you a way to have functioning tabs out of the box, why can't we use it?

Reply:
Because I do not want to provide templates, I need to know which tab is opened to render the fields described in the map.

Reply:
But the ui.bootstrap tabs support dynamic content out of the box, what's the technical limitation? Also, in your solution you are basically recreating/destroying the form of each tab any time get selected/deselected, i think that's something we should avoid performance-wise

Reply:
UIBootstrap will render all fields at once and then just toggle the tab contents - that's right. On the other hand, imagine the wizard is large and contains a lot of tabs and fields, it is better to render on demand. The rendering in that case will take complexity O(fields_amount_per_tab) while in the case of tabs it will be O(all_fields_from_all_tabs).

As per inventing a wheel I will still need to toggle `active` class as described here in the example https://angular-ui.github.io/bootstrap/. I do not see a need of use UI bootstrap here, the solution is clean enough as per my opinion and it uses bootstrap markup.

Reply:
> UIBootstrap will render all fields at once and then just toggle the tab contents - that's right. On the other hand, imagine the wizard is large and contains a lot of tabs and fields, it is better to render on demand. The rendering in that case will take complexity O(fields_amount_per_tab) while in the case of tabs it will be O(all_fields_from_all_tabs).

On hand we are talking about displaying a bunch of DOM nodes from the start, which is something that browsers can handle pretty well. On the other we are talking about doing some heavy DOM manipulation on each user interaction via angular, which we know from past experience is something we should avoid if we can

> As per inventing a wheel I will still need to toggle active class as described here in the example https://angular-ui.github.io/bootstrap/. I do not see a need of use UI bootstrap here, the solution is clean enough as per my opinion and it uses bootstrap markup.

I'm not questioning the elegance of your solution, but if we are paying the cost of including a library in our codebase, then we should try to use what that library is providing to us instead of writing a solution ourselves.

Unless there are technical reasons for why using the tabs directive provided by ui.bootstrap are limiting or hindering your solution (and there might be, for example with how the directive expects to write the markup), I don't see a reason for us not to use it. Please don't write code if it's not necessary

Reply:
I attempted to do my best.